File: app/src/main/java/com/ocram/qichwadic/framework/net/client/RetrofitService.java
package com.ocram.qichwadic.framework.net.client;

import com.ocram.qichwadic.BuildConfig;
import com.ocram.qichwadic.domain.model.Definition;
import com.ocram.qichwadic.domain.model.Dictionary;
import com.ocram.qichwadic.domain.model.SearchResult;

import java.util.List;

import io.reactivex.Flowable;
import retrofit2.http.GET;
import retrofit2.http.Path;
import retrofit2.http.Query;

public interface RetrofitService {

    String API_URL = BuildConfig.SERVER_URL;
    String ALL_DICTIONARIES = API_URL + "dictionaries";
    String SEARCH_WORDS = API_URL + "words";

    @GET(ALL_DICTIONARIES)
    Flowable<List<Dictionary>> getDictionaries(@Query("all") int all);

    @GET(SEARCH_WORDS)
    Flowable<List<SearchResult>> searchWord(
            @Query("all") int all,
            @Query("searchWord") String searchWord,
            @Query("fromQuechua") int fromQuechua,
            @Query("target") String target,
            @Query("searchType") int searchType
    );

    @GET(ALL_DICTIONARIES + "/{dictionaryId}/words/fetch")
    Flowable<List<Definition>> fetchMoreResults(
            @Path("dictionaryId") int dictionaryId,
            @Query("searchWord") String searchWord,
            @Query("fromQuechua") int fromQuechua,
            @Query("searchType") int searchType,
            @Query("page") int page
    );

    @GET(ALL_DICTIONARIES + "/{dictionaryId}/words")
    Flowable<List<Definition>> getAllDefinitionsByDictionary(@Path("dictionaryId") int dictionaryId);
}

File: app/src/main/java/com/ocram/qichwadic/features/common/data/remote/RetrofitService.kt
package com.ocram.qichwadic.features.common.data.remote

import com.ocram.qichwadic.BuildConfig
import com.ocram.qichwadic.features.common.data.model.DefinitionEntity
import com.ocram.qichwadic.features.common.data.model.DictionaryEntity
import com.ocram.qichwadic.features.common.data.model.SearchResultEntity
import retrofit2.Response

import retrofit2.http.GET
import retrofit2.http.Path
import retrofit2.http.Query

interface RetrofitService {

    @GET(ALL_DICTIONARIES)
    suspend fun getDictionaries(@Query("all") all: Int): Response<List<DictionaryEntity>>

    @GET(SEARCH_WORDS)
    suspend fun searchWord(
            @Query("all") all: Int,
            @Query("searchWord") searchWord: String,
            @Query("fromQuechua") fromQuechua: Int,
            @Query("target") target: String,
            @Query("searchType") searchType: Int
    ): Response<List<SearchResultEntity>>

    @GET("$ALL_DICTIONARIES/{dictionaryId}/words/fetch")
    suspend fun fetchMoreResults(
            @Path("dictionaryId") dictionaryId: Int,
            @Query("searchWord") searchWord: String,
            @Query("fromQuechua") fromQuechua: Int,
            @Query("searchType") searchType: Int,
            @Query("page") page: Int
    ): Response<SearchResultEntity>

    @GET("$ALL_DICTIONARIES/{dictionaryId}/words")
    suspend fun getAllDefinitionsByDictionary(@Path("dictionaryId") dictionaryId: Int): Response<List<DefinitionEntity>>

    companion object {

        const val API_URL = BuildConfig.SERVER_URL
        const val ALL_DICTIONARIES = API_URL + "dictionaries"
        const val SEARCH_WORDS = API_URL + "words"
    }
}
